transport_plugins/websocket/iotile_transport_websocket/device_adapter.py

```diff
@@ -102,6 +155,281 @@ def stop_sync(self):
+            return
+
+        self.connections.begin_operation(connection_id, 'rpc', callback, timeout)
+
+        try:
+            result = self.client.send_command('send_rpc', {
+                'address': address,
+                'rpc_id': rpc_id,
+                'payload': payload,
+                'timeout': timeout
+            })
+            messages.RPCResponse.verify(result)
+        except ValidationError as err:
+            reason = "Wrong response received from ws server after sending 'send_rpc' command: {}". format(err)
+            self.connections.finish_operation(connection_id, False, reason, None, None)
+            raise HardwareError(reason)
+        except Exception as err:
+            reason = "Error while sending 'send_rpc' command to ws server: {}".format(err)
+            self.connections.finish_operation(connection_id, False, reason, None, None)
+            raise HardwareError(reason)
+
+        payload = result.get('payload')
+        status = 0xFF  # Default value, meaning that an error occurred
+        return_value = bytearray()
+
+        if payload is not None:
+            status = payload.get('status', status)
+            return_value = payload.get('return_value', return_value)
+
+        # FIXME: async after creating WebSocketHandler2
+        self.connections.finish_operation(
+            connection_id,
+            result['success'],
+            result.get('reason', None),
+            status,
+            return_value
+        )
+
+    def send_script_async(self, connection_id, data, progress_callback, callback):
+        """Asynchronously send a a script to this IOTile device
+
+        Args:
+            connection_id (int): A unique identifier that will refer to this connection
+            data (string): the script to send to the device
+            progress_callback (callable): A function to be called with status on our progress, called as:
+                progress_callback(done_count, total_count)
+            callback (callable): A callback for when we have finished sending the script. The callback will be called as
+                callback(connection_id, adapter_id, success, failure_reason)
+                'connection_id': the connection id
+                'adapter_id': this adapter's id
+                'success': a bool indicating whether we received a response to our attempted RPC
+                'failure_reason': a string with the reason for the failure if success == False
+        """
+
+        try:
+            context = self.connections.get_context(connection_id)
+        except ArgumentError:
+            callback(connection_id, self.id, False, "Could not find connection information")
+            return
+
+        context['progress_callback'] = progress_callback
+
+        self.connections.begin_operation(connection_id, 'script', callback, self.get_config('default_timeout'))
+
+        # Count number of chunks to send
+        nb_chunks = 1
+        if len(data) > self.mtu:
```

Comment:
I would not cache `self.mtu` but rather request it from `get_config` once per function that needs it since otherwise there is no opportunity for anyone to override the default value because it gets cached into `self.mtu` in `__init__` and self.mtu is never updated again.